### ISSUE TITLE
S3Store successfully saving info but not bin file

### ISSUE THREAD

**Issue by amlwwalker** [tus/tusd#252]
Hi, I'm using tusd v1 in to create a client and server to upload files directly to s3 in multiparts, thereby not saving a temporary file to the server in the mean time.

It seems to be generally working, however I suddenly noticed its not saving the .bin files (or atleast the file that I actually want) to s3 its just saving the .info files.

The [code for the client and the server are here](https://gist.github.com/amlwwalker/8424b849b65f0c098b9174725b90d315), I thought best to provide a full source. Hopefully I have left a parameter out or something. 
Help greatly appreciated.

S3 Looks like the below image. As you can see its having no problems writing to S3 it just seems to not be writing the actual file.

_Incidently as you can see in my code, I am using regex to extract the name of the file when it's uploaded so that I can store that in a db and request it at a later date. Is this the correct way, or is there a neater way?_

![image](https://user-images.githubusercontent.com/2920180/54479560-d6fe1680-4815-11e9-85e8-b588a8c7fdc0.png)

**Comment by Acconut** [comment]
S3Store does not create a `.bin` file on S3 as it does for the filestore. Instead it creates a multipart upload and stores the data in this multipart upload. However, you cannot view this multipart upload in the dashboard while the upload is not finished. So once the upload is done you should see the complete objects in S3. For more details, this comment explained the storage on S3 a bit more: https://github.com/tus/tusd/issues/244#issuecomment-468447969

**Comment by Acconut** [comment]
> Incidently as you can see in my code, I am using regex to extract the name of the file when it's uploaded so that I can store that in a db and request it at a later date. Is this the correct way, or is there a neater way?

No, unfortunately there is currently no better way.

**Comment by amlwwalker** [comment]
oh wow, right (thanks for getting back to me)
So, that means my uploads aren't completing...?
What could be causing that? I have tried to wrap the tusd.NewHandler inside another handler in my server

```
func CustomHandler(tusdHandler *tusd.Handler, profile *models.GoogleProfile, composer *tusd.StoreComposer) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 tusdHandler.ServeHTTP(w, r)
   }
}
```

passing the `tusdHandler` in to that and then seving from there so that I can set the file/bytes etc before tusd gets a hold of it, but I am wondering if there is something wrong with doing it like that. Do you know if this would work?
Basically the reason is the data to write to s3 is raw bytes rather than a file, coming in over json. I don't suppose that will cause any problems? I cant find a reference to what tusd is looking for to write to s3.... (content type etc...) or does it just take whatever is in the body of the request and send that?

**Comment by Acconut** [comment]
> If the client (the thing requesting the upload) is not using multipart to send the data, will tusd still be able to write it? I guessed it would just manage the data as they came in....

I think you are confusing two things here: The AWS S3 multipart uploads (https://docs.aws.amazon.com/AmazonS3/latest/dev/mpuoverview.html) are entirely unrelated to the HTTP multipart content type. While both are usually called "multipart upload" they are two very things depending on what you are talking about.

Now, tusd uses the AWS S3 multipart uploads internally for uploading the data to S3 but this is never exposed to the client.

Furthermore, tus does not use the HTTP multipart content type at all. So using Go's mime/multipart package for uploading to tusd is wrong. From looking at your code it seems as if you didn't understand how tus works. Let me quickly explain it by answer your other question:

> So, that means my uploads aren't completing...?
What could be causing that?
I am getting a 201 response from tusd which makes me believe its completed the write. I'm not getting any errors anywhere, can I use the hooks in anyway to monitor progress?

In tus there is a difference between creating an upload and uploading the actual file. First you create a new empty upload using the POST request and get a 201 response with the Location header pointing to your upload URL. This POST request does only transfer metadata, such as the filesize, but not the actual file itself. The 201 code you are seeing is the response for the POST request.

After that you send a PATCH request to the upload URL you just got back. This PATCH request contains the file you want to upload, but it does not use the HTTP multipart content type! This request is entirely missing from your code, so it won't work at all. I recommend you read the specification to understand how does requests need to be performed: https://tus.io/protocols/resumable-upload.html

**Comment by amlwwalker** [comment]
Ok thanks, I'll have a read...

I was playing around, and the following curl script works with my server code (I get the file).

```
file=${1}
filename=$(basename "${file}" | base64)
filesize="$(wc -c <"${file}")"

echo "file size" $filesize
curl \
  -X POST \
  -H "Tus-Resumable: 1.0.0" \
  -H "Upload-Offset: 0" \
  -H "Content-Length: ${filesize}" \
  -H "Upload-Length: ${filesize}" \
  -H "Content-Type: application/offset+octet-stream" \
  --data-binary "@${file}" \
  http://localhost:8080/files/upload/ -vvv
```

This possibly has confused me as that seems to just set the content type (agreed, not to multipart, however I have now tried it aswell).
So is curl here following the redirect and then doing the PATCH in the background?

**Comment by amlwwalker** [comment]
EDIT:
I have now got it POSTING (not HEAD as it seems to not have the location header) and I get the 201 back and I can see the .info file in s3.
So then I PATCH:

```
reqPatch, err := http.NewRequest("PATCH", postResponse.Header["Location"][0], body)

	if err != nil {
		log.Fatal(err)
	}

	fmt.Printf("The file is %s bytes long\r\n", fileSize)
	reqPatch.Header.Add("Tus-Resumable", "1.0.0")
	reqPatch.Header.Add("Upload-Offset", "0")
	reqPatch.Header.Add("Content-Length", fileSize)
	reqPatch.Header.Add("Upload-Length", fileSize)
	reqPatch.Header.Add("Content-Type", "application/offset+octet-stream")

	fmt.Printf("patch headers %+v\r\n", reqPatch.Header)
```

from that I get

```
[tusd] event="ResponseOutgoing" status="400" method="PATCH" path="6c16df183933e984b7d1b61ba53d958b+Ydk2hfJjTirxfkR60bpfW0xaBekbKVmSjJra7kKW1r9eRdl8BidN384TAfHRwfqKftjopA4Ruc26LyVnKAvpKrYRfEMLrdisbJNAkUrttPJLB9vMpn8teAgiPgK9zrNR" error="missing or invalid Upload-Length header"
```
But when I print the headers, I can see I do have the `Upload-Length` header

```
patch headers map[Tus-Resumable:[1.0.0] Upload-Offset:[0] Content-Length:[1566] Upload-Length:[1566] Content-Type:[application/offset+octet-stream]]
```

I cant quite visualise the flow here. Do I have to manually loop over the PATCH request, continually updating the offset and sending the next chunk, or does that happen automagically and the offset is just to start off a download again in the case where it stops for some reason?

**Comment by amlwwalker** [comment]
OK I have it working now using a library I found for client side interaction with tus.

Incidently, I mentioned I had the tusd handler inside a wrapper handler in a previous message. I am using the wrapper to check for authentication (an API key as a query parameter). This all works, but it causes tus to continuously loop for ever when uploading the chunks, I think possibly because the offset may not be being updated, but i thought that was only if the connection was restarted. I am wondering if it is not being passed the headers or something, have you any experience with what could cause this?

**Comment by Acconut** [comment]
> This possibly has confused me as that seems to just set the content type (agreed, not to multipart, however I have now tried it aswell).
So is curl here following the redirect and then doing the PATCH in the background?

No, curl is not performing two requests under the hood. However, tusd implements the (not yet standardized) "Creation with Upload" extension (see https://github.com/tus/tus-resumable-upload-protocol/blob/0ed3a96dd8006cf6148716fd4b0ea556fd9790b1/protocol.md#creation-with-upload). It basically allows you to combine the POST and PATCH request into one single request. However, since most tus servers do not support this feature and it's intended for advanced users, I didn't mention it.

> from that I get
> [tusd] event="ResponseOutgoing" status="400" method="PATCH" path="6c16df183933e984b7d1b61ba53d958b+Y..." error="missing or invalid Upload-Length header"

Your PATCH request should not contain the Upload-Length header at all. After removing it, it should work. Again, this would also be covered in the protocol specification.

> Do I have to manually loop over the PATCH request, continually updating the offset and sending the next chunk

Yes, for every PATCH request you need to set the correct offset. This is to ensure that the client and server are storing the chunk at the correct position.

> Incidently, I mentioned I had the tusd handler inside a wrapper handler in a previous message. I am using the wrapper to check for authentication (an API key as a query parameter). This all works, but it causes tus to continuously loop for ever when uploading the chunks, I think possibly because the offset may not be being updated, but i thought that was only if the connection was restarted. I am wondering if it is not being passed the headers or something, have you any experience with what could cause this?

I have no idea. Without more details, I won't be able to help you.

**Comment by amlwwalker** [comment]
Thanks for that, makes sense, and yes following along your comments with the documentation, probably just me, but sometimes documentation alone can feel a little abstract, your clarification helps!

So I'm using the beego web framework, and have incorporated the tusd handler within that.

If my handler is

```
const UPLOAD_PATH = /files/upload/

	handler, err := tusd.NewHandler(tusd.Config{
		BasePath:      UPLOAD_PATH,
		StoreComposer: composer,
	})
	if err != nil {
		panic(fmt.Errorf("Unable to create handler: %s", err))
	}
```
and I incorporate it as the endpoint handler with beego framework with:
`beego.Handler(UPLOAD_PATH, http.StripPrefix(UPLOAD_PATH, handler), true)`

everything works as expected. However I want to access some of the information that is sent over the request before it gets sent to s3. The object I am writing is infact json, so I want one of the fields out of the json. To do that, i am (trying to) wrapping the tusd handler in another HTTP handler, so I change the `beego.Handler` line above to 

```
beego.Handler(UPLOAD_PATH, http.StripPrefix(UPLOAD_PATH, CustomHandler(handler)), true)
```

and then customHandler looks like

```
func CustomHandler(tusdHandler *tusd.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		_, err := ioutil.ReadAll(r.Body)
		if err != nil {
			log.Printf("Error reading body: %v", err)
			return
		}
		tusdHandler.Handler.ServeHTTP(w, r)
	})
}
```

If I do it directly (the first case) it works perfectly. If I try and intercept the request and read the data off in the handler wrapper, it causes it to never complete the process and I get these lines looping forever:

```
[tusd] event="RequestIncoming" method="PATCH" path="482bcf22aa3f8ec1ba2a6011631a1305+SdT2pCMKTci6SyVRyvvRwnbuSDZzpkac3_dDWMU671R2fkR1V9uqQwfab5rst4w1VWEhG.LViDOLQaUL7Ve4duEbMnCCJKzSb_ZzIn3U0HdJlHdiEe1iLcQxAdXnT29x"
[tusd] event="ChunkWriteStart" id="482bcf22aa3f8ec1ba2a6011631a1305+SdT2pCMKTci6SyVRyvvRwnbuSDZzpkac3_dDWMU671R2fkR1V9uqQwfab5rst4w1VWEhG.LViDOLQaUL7Ve4duEbMnCCJKzSb_ZzIn3U0HdJlHdiEe1iLcQxAdXnT29x" maxSize="1566" offset="0"
[tusd] event="ChunkWriteComplete" id="482bcf22aa3f8ec1ba2a6011631a1305+SdT2pCMKTci6SyVRyvvRwnbuSDZzpkac3_dDWMU671R2fkR1V9uqQwfab5rst4w1VWEhG.LViDOLQaUL7Ve4duEbMnCCJKzSb_ZzIn3U0HdJlHdiEe1iLcQxAdXnT29x" bytesWritten="0"
[tusd] event="ResponseOutgoing" status="204" method="PATCH" path="482bcf22aa3f8ec1ba2a6011631a1305+SdT2pCMKTci6SyVRyvvRwnbuSDZzpkac3_dDWMU671R2fkR1V9uqQwfab5rst4w1VWEhG.LViDOLQaUL7Ve4duEbMnCCJKzSb_ZzIn3U0HdJlHdiEe1iLcQxAdXnT29x"
```

As opposed to the 204 completing and that be that.
Any knowledge or suggestions would be really helpful, thanks

**Comment by Acconut** [comment]
Closing this issue due to inactivity. Feel free to leave a comment if you want to continue the discussion :)
